### PR TITLE
Make SearchTemplateRequest implement IndicesRequest.Replaceable

### PR DESCRIPTION
**Pull request by cwperks**
### Description

Companion PR in Security plugin that includes tests of the behavior: https://github.com/opensearch-project/security/pull/2921

This PR makes SearchTemplateRequest (inside `modules/lang-mustache`) implement `IndicesRequest.Replaceable` and for the implementation of each method it calls the underlying SearchRequest's implementation. 

This change resolves an issue in authorization of SearchTemplateRequests where the security plugin is unable to extract the indices associated with the request properly. See Details: https://github.com/opensearch-project/security/issues/1678

As an effect, even if a user does a SearchTemplateRequest on a specific index:

i.e.

```
GET /movies/_search/template
{
  "source": {
    "query": {
      "match": {
        "title": "{{title}}"
      }
    }
  },
  "params": {
    "title": "Titanic"
  }
}
```

within the security plugin, it is not able to determine the request is on `movies` index and instead tries to evaluate whether the user has permission to do search template requests on `*` index. 

With this change, the Security plugin will be able to extract the indices from the request as is already done for `Replaceable` requests here: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L752-L758

### Related Issues

Resolves https://github.com/opensearch-project/security/issues/1678

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
Are any maintainers available for a PR review?

**Comment by cwperks** [comment]
Tests for this are being added in the security plugin here: https://github.com/opensearch-project/security/pull/2921/files

**Comment by cwperks** [comment]
@reta Could I trouble you for a review of this PR? It's a small one for a bug fix related to privilege evaluation of SearchTemplateRequests

**Comment by cwperks** [comment]
@reta or @dblock Would either of you be able to review this PR? This calls on the underlying implementations in the SearchRequest that a part of the SearchTemplateRequest. This fixes an issue with how SearchTemplateRequests are authorized where the security plugin cannot currently get the list of indices the SearchTemplateRequest is executed on and instead assumes `*` meaning that a user needs a read permission on `*` index pattern to do any SearchTemplateRequests.

**Comment by reta** [comment]
> @reta or @dblock Would either of you be able to review this PR?

Sorry about the delay @cwperks , looking right now

File: modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java
Comment (reta) [inline modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java:61]: @cwperks why do you need replaceable here? From the description, the issue is that indices could not be extracted, in this case `implements IndicesRequest` is fully sufficient,

Comment (cwperks) [reply modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java:61]: Also tagging @peternied

Comment (cwperks) [reply modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java:61]: If I'm understanding the flow in IndexReplacerResolver, the replacements only come into the picture when dnfof is enabled.

dnfof === do not fail on forbidden. Looking into that more now to see if that's applicable for these requests too.

Comment (cwperks) [reply modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java:61]: Oh I get it now. With `dnfof` enabled the security plugin will go through all roles mapped to a user and compare the indices from the request against with the list of indices the user is permitted to for the given action. The security plugin modifies the request to only include the indices from the request that the user is permitted to for the given action.

For instance let's say I have user `craig` that is permitted to do `indices:data/read/search*` on `logs`, but not on `movies`. 

1. With dnfof disabled, the follow request would fail because `craig` does not have permission to search `movies` index even though it can search `logs` index:

```
curl -XGET https://craig:TestPassword123_@localhost:9200/logs,movies/_search/template --insecure -H "Content-Type: application/json" --data '
{
  "source": {
    "query": {
      "match": {
        "title": "{{title}}"
      }
    }
  },
  "params": {
    "title": "Titanic"
  }
}
'
```

2. With dnfof enabled the request would work, but it would only in effect to the query against `logs` index and the logic in the security plugin replaces the index before the search is performed.

Comment (cwperks) [reply modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java:61]: @reta ^

**Review by peternied — CHANGES_REQUESTED** [review]
Search request templates do not not fit in the security plugins model, they are cluster wide construction.  I don't think we should fix https://github.com/opensearch-project/security/issues/1678 which is driving this change.

Seeing aside the merits of that bug - the notion of a replaceable indices on request should be the standard for plugin capabilities.  If we are taking this change from an platform perspective should `IndicesRequest.Replaceable` be more prevalent (including this case)?  @macohen can you help us find someone to advise on next steps?

**Review by peternied — APPROVED** [review]
Dug in on https://github.com/opensearch-project/security/issues/1678 the intent is around fixing a place where `GET /*/_search` works as expected and `GET /*/_search/template` is broken, so lets unblock that scenario with this change.

Thanks @cwperks

**Review by reta — APPROVED** [review]

File: modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java
Comment (nibix) [reply modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateRequest.java:61]: @cwperks 

> @nibix I know you mentioned a bit about adding `IndicesRequest` into `IndexResolverReplacer.getOrReplaceAllIndices` before, but I see its still not in the big `if { } else { }` block. Do you know if there is a good reason why it is not handled already because `Replaceable` is.

Sorry, I am a bit late to the game. Generally, I do not think there is a good reason why IndicesRequest is not handled here. Most likely it was just not needed so far. 

Just as a BTW/FYI, another place in the security plugin which utilizes the Replaceable interface is the PrivilegesInterceptor:

https://github.com/opensearch-project/security/blob/0d915e24cf16913f01d7d97a9a007c99b39bc2ed/src/main/java/org/opensearch/security/configuration/PrivilegesInterceptorImpl.java#L349

**Pull request merged**
